--- oss-v4.2-build2011-src-gpl/configure	2014-02-14 15:59:55.000000000 -0700
+++ oss-v4.2-build2011-src-gpl-fix/configure	2015-12-03 06:02:16.860029598 -0700
@@ -13,8 +13,8 @@
 #
 # Set GRC3 min/max conversion quality setting to 3 to conserve space.
 # TODO: Add support for configurable quality settings.
-GRC_MIN_QUALITY=3
-GRC_MAX_QUALITY=3
+GRC_MIN_QUALITY=6
+GRC_MAX_QUALITY=6
 export GRC_MIN_QUALITY GRC_MAX_QUALITY
 
 # MIDI support is enabled by default. Use --config-midi=NO to disable
